File: Week 2 -- mixed-effects/Homework/normInvGamma.R
library(TMB)
library(data.table)
library(MCMCpack)
library(ggplot2)

# Data Generation ---------------------------------------------------------

#' Empirical Bayes estimation of Normal likelihood, normal prior on mu, log-normal prior on sigma

sim_data <- function(n, J, mu0, sigma_mu, l_mu_sigma, l_sigma_sigma){
  #' sigma_s   ~ LogN(l_mu_sigma, l_sigma_sigma)
  #' mu_s      ~ N(mu_0, sigma_mu)
  #' Y         ~ N(mu_s, sigma_s)
  #' 
  #' Fixed: mu_0, sigma_mu, l_mu_sigma, l_sigma_sigma
  #' Random: sigma_s, mu_s
  
  #' Group labels of the data  
  s_i <- sample(1:J, n, replace=TRUE)
  
  #' site specific means
  mu_s <- rnorm(J, mu0, sigma_mu)
  
  #' site specific variance
  log_sigma_s <- rnorm(J, l_mu_sigma, l_sigma_sigma)
  
  #' Observed data
  y_i <- sapply(s_i, function(j) rnorm(1, mu_s[j], exp(log_sigma_s[j])))
  
  df <- data.table(y_i=y_i, s_i=s_i)
  return(df)
}

df <- sim_data(1000, 10, 10, 3, 1, .5)

ggplot(df, aes(as.factor(s_i), y_i)) + geom_boxplot()

# TMB ---------------------------------------------------------------------

#' TMB only works well with normal priors...
#' Wanted to do InvGamma on variance, but maybe will try LogNormal

Data <- list( "n" = nrow(df), 
              "n_s" = length(unique(df$s_i)), 
              "y_i" = df$y_i, 
              "s_i" = df$s_i-1 )

#' l_mu_sigma: mean of log(sigma_s)
#' l_sigma_sigma: sigma of log(sigma_s)
Parameters <- list( "mu_0" = 0,
                    "log_sigma_mu" = 0,
                    "l_mu_sigma" = 0, 
                    "l_sigma_sigma" = 1, 
                    "mu_s" = rep(0, Data$n_s), 
                    "log_sigma_s" = rep(0, Data$n_s))

Random <- c("mu_s", "log_sigma_s")

compile("normInvGamma.cpp")
dyn.load( dynlib("normInvGamma"))

Obj <- MakeADFun(data = Data, parameters = Parameters, random = Random)
Obj$env$beSilent()
Opt <- TMBhelper::fit_tmb(Obj)

sd_table <- summary(sdreport(Obj))[c('mu_0', 'sigma_mu', 'l_mu_sigma', 'l_sigma_sigma'), ]
sd_table

File: Week 2 -- mixed-effects/Homework/normInvGamma.cpp
#include <TMB.hpp>
template <class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER( n );
  DATA_INTEGER( n_s );
  DATA_VECTOR( y_i );
  DATA_IVECTOR( s_i );
  
  //Fixed effects estimate by EB
  PARAMETER( mu_0 );
  PARAMETER( log_sigma_mu );
  PARAMETER( l_mu_sigma );
  PARAMETER( l_sigma_sigma );
  
  //Random effects
  PARAMETER_VECTOR( mu_s );
  PARAMETER_VECTOR( log_sigma_s );
  
  Type jnll=0;
  
  // y_i data likelihood
  for(int i=0; i<n; i++) {
    jnll -= dnorm( y_i(i), mu_s(s_i(i)), exp( log_sigma_s(s_i(i)) ), true );
  }
  
  // mu_s random likelihood
  for(int j=0; j<n_s; j++){
    jnll -= dnorm( mu_s(j), mu_0, exp(log_sigma_mu), true );
  }
  
  // log_sigma_s random likelihood
  for(int j=0; j<n_s; j++){
    jnll -= dnorm( log_sigma_s(j), l_mu_sigma, l_sigma_sigma, true );
  }
  
  Type sigma_mu = exp(log_sigma_mu);
  vector<Type> sigma_s = exp(log_sigma_s);
  
  ADREPORT( mu_0 );
  ADREPORT( sigma_mu );
  ADREPORT( l_mu_sigma );
  ADREPORT( l_sigma_sigma );
  ADREPORT( mu_s );
  ADREPORT( sigma_s );
  
  REPORT( mu_0 );
  REPORT( sigma_mu );
  REPORT( l_mu_sigma );
  REPORT( l_sigma_sigma );
  REPORT( mu_s );
  REPORT( sigma_s );
  
  return jnll;
}